### ISSUE TITLE
Cannot add missing static method to final API through additions

### ISSUE THREAD

**Issue by LeoJHarris** [xamarin/xamarin-android#5825]
Hi, I am trying to add a static method into an existing class `Preconditions` in my binding library through the `additions` class, the method is required and called from somewhere else in the native android library SDK. The method I am trying to add is this https://github.com/google/dagger/blob/ba49e287142deb322b5bcd20ff4a9aee7b49f1ac/java/dagger/internal/Preconditions.java#L60 as also shown below:

```csharp
/**
   * Ensures that an object reference returned from a provides method is not null.
   *
   * @param reference an object reference
   * @return the non-null reference that was validated
   * @throws NullPointerException if {@code reference} is null
   */
  public static <T> T checkNotNullFromProvides(T reference) {
    if (reference == null) {
      throw new NullPointerException("Cannot return null from a non-@Nullable @Provides method");
    }
    return reference;
  }
```

Long story short this method exists in a latest recent version of dagger but I NEED an older version of dagger 2.27.0 and unfortunately this method was not present in that version so I am trying to add this missing method myself but its not working but I know exactly how the method should be defined and have implemented it in additions, the below would appear to me to be the correct way to add this method in the final API of the binding library but its not working

My Additions looks like this:

```csharp
namespace Dagger.Internal
{
    partial class Preconditions : global::Java.Lang.Object
    {
        [Register("My", "(Ljava/lang/Object;Ljava/lang/Class;)V", "")]
        public static T CheckNotNullFromProvides<T>(T reference)
        {
            return reference ?? throw new NullReferenceException("Cannot return null from a non-@Nullable @Provides method");
        }
    }
...
}
```

The generated Dagger.Internal.Preconditions.cs also looks like this:

![Capture](https://user-images.githubusercontent.com/5468757/113941238-d5677700-9852-11eb-99a0-652cdad43d35.PNG)

I also looked in the api.xml and couldn't find the method I tried to add and when I run the application it still throws exception of missing static method:

```csharp
[AndroidRuntime] FATAL EXCEPTION: main
[AndroidRuntime] Process: sampleapp.com.au, PID: 26685
[AndroidRuntime] java.lang.NoSuchMethodError: No static method checkNotNullFromProvides(Ljava/lang/Object;)Ljava/lang/Object; in class Ldagger/internal/Preconditions; or its super classes (declaration of 'dagger.internal.Preconditions' appears in /data/app/sampleapp.com.au-V8X3AEguYW91QI05nanQ6A==/base.apk)
[AndroidRuntime] 	at sqip.internal.HttpModule_MoshiFactory.moshi(HttpModule_MoshiFactory.java:23)
[AndroidRuntime] 	at sqip.internal.HttpModule_MoshiFactory.get(HttpModule_MoshiFactory.java:15)
[AndroidRuntime] 	at sqip.internal.HttpModule_MoshiFactory.get(HttpModule_MoshiFactory.java:8)
[AndroidRuntime] 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
[AndroidRuntime] 	at sqip.internal.HttpModule_CreateCardNonceErrorResponseAdapterFactory.get(HttpModule_CreateCardNonceErrorResponseAdapterFactory.java:24)
[AndroidRuntime] 	at sqip.internal.HttpModule_CreateCardNonceErrorResponseAdapterFactory.get(HttpModule_CreateCardNonceErrorResponseAdapterFactory.java:11)
[AndroidRuntime] 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
[AndroidRuntime] 	at sqip.internal.nonce.RealCreateCardNonceRequestHandler_Factory.get(RealCreateCardNonceRequestHandler_Factory.java:44)
[AndroidRuntime] 	at sqip.internal.nonce.RealCreateCardNonceRequestHandler_Factory.get(RealCreateCardNonceRequestHandler_Factory.java:12)
[AndroidRuntime] 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
[AndroidRuntime] 	at sqip.internal.nonce.ActivityControllerModule_CardEntryActivityControllerFactory.get(ActivityControllerModule_CardEntryActivityControllerFactory.java:36)
[AndroidRuntime] 	at sqip.internal.nonce.ActivityControllerModule_CardEntryActivityControllerFactory.get(ActivityControllerModule_CardEntryActivityControllerFactory.java:11)
[AndroidRuntime] 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
[AndroidRuntime] 	at sqip.internal.nonce.DaggerCardEntryActivityComponent.cardEntryActivityController(DaggerCardEntryActivityComponent.java:160)
[AndroidRuntime] 	at sqip.internal.BaseCardEntryActivity.onCreate(BaseCardEntryActivity.kt:132)
[AndroidRuntime] 	at sqip.internal.CardEntryActivity.onCreate(CardEntryActivity.kt:38)
[AndroidRuntime] 	at android.app.Activity.performCreate(Activity.java:7136)
[AndroidRuntime] 	at android.app.Activity.performCreate(Activity.java:7127)
[AndroidRuntime] 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1271)
[AndroidRuntime] 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2893)
[AndroidRuntime] 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3048)
[AndroidRuntime] 	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
[AndroidRuntime] 	04-08 11:34:59.852 E/AndroidRuntime(26685): 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
[AndroidRuntime] 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
[AndroidRuntime] 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
[AndroidRuntime] 	at android.os.Handler.dispatchMessage(Handler.java:106)
[AndroidRuntime] 	at android.os.Looper.loop(Looper.java:193)
[AndroidRuntime] 	at android.app.ActivityThread.main(ActivityThread.java:6669)
[AndroidRuntime] 	at java.lang.reflect.Method.invoke(Native Method)
[AndroidRuntime] 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
[AndroidRuntime] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
**Java.Lang.NoSuchMethodError:** 'No static method checkNotNullFromProvides(Ljava/lang/Object;)Ljava/lang/Object; in class Ldagger/internal/Preconditions; or its super classes (declaration of 'dagger.internal.Preconditions' appears in /data/app/sampleapp.com.au-V8X3AEguYW91QI05nanQ6A==/base.apk)'

[Mono] DllImport attempting to load: '/system/lib/liblog.so'.
[Mono] DllImport loaded library '/system/lib/liblog.so'.
[Mono] DllImport searching in: '/system/lib/liblog.so' ('/system/lib/liblog.so').
[Mono] Searching for '__android_log_print'.
[Mono] Probing '__android_log_print'.
[Mono] Found as '__android_log_print'.
[MonoDroid] UNHANDLED EXCEPTION:
[MonoDroid] Java.Lang.NoSuchMethodError: No static method checkNotNullFromProvides(Ljava/lang/Object;)Ljava/lang/Object; in class Ldagger/internal/Preconditions; or its super classes (declaration of 'dagger.internal.Preconditions' appears in /data/app/sampleapp.com.au-V8X3AEguYW91QI05nanQ6A==/base.apk)
[MonoDroid]   --- End of managed Java.Lang.NoSuchMethodError stack trace ---
[MonoDroid] java.lang.NoSuchMethodError: No static method checkNotNullFromProvides(Ljava/lang/Object;)Ljava/lang/Object; in class Ldagger/internal/Preconditions; or its super classes (declaration of 'dagger.internal.Preconditions' appears in /data/app/sampleapp.com.au-V8X3AEguYW91QI05nanQ6A==/base.apk)
[MonoDroid] 	at sqip.internal.HttpModule_MoshiFactory.moshi(HttpModule_MoshiFactory.java:23)
[MonoDroid] 	at sqip.internal.HttpModule_MoshiFactory.get(HttpModule_MoshiFactory.java:15)
[MonoDroid] 	at sqip.internal.HttpModule_MoshiFactory.get(HttpModule_MoshiFactory.java:8)
[MonoDroid] 	at sqip.internal.HttpModule_MoshiFactory.get(HttpModule_MoshiFactory.java:04-08 11:35:03.580 I/MonoDroid(26685): 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
[MonoDroid] 	at sqip.internal.HttpModule_CreateCardNonceErrorResponseAdapterFactory.get(HttpModule_CreateCardNonceErrorResponseAdapterFactory.java:24)
[MonoDroid] 	at sqip.internal.HttpModule_CreateCardNonceErrorResponseAdapterFactory.get(HttpModule_CreateCardNonceErrorResponseAdapterFactory.java:11)
[MonoDroid] 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
[MonoDroid] 	at sqip.internal.nonce.RealCreateCardNonceRequestHandler_Factory.get(RealCreateCardNonceRequestHandler_Factory.java:44)
[MonoDroid] 	at sqip.internal.nonce.RealCreateCardNonceRequestHandler_Factory.get(RealCreateCardNonceRequestHandler_Factory.java:12)
[MonoDroid] 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
[MonoDroid] 	at sqip.internal.nonce.ActivityControllerModule_CardEntryActivityControllerFactory.get(ActivityControllerModule_CardEntryActivityControllerFactory.java:36)
[MonoDroid] 	at sqip.internal.nonce.ActivityControllerModule_CardEntryActivityControllerFactory.get(ActivityControllerModule_CardEntryActivityControllerFactory.java:11)
[MonoDroid] 	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
[MonoDroid] 	at sqip.internal.nonce.DaggerCardEntryActivityComponent.cardEntryActivityController(DaggerCardEntryActivityComponent.java:160)
[MonoDroid] 	at sqip.internal.CardEntryActivity.onCreate(CardEntryActivity.kt:38)
[MonoDroid] 	at android.app.Activity.performCreate(Activity.java:7136)
[MonoDroid] 	at android.app.Activity.performCreate(Activity.java:7127)
[MonoDroid] 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1271)
[MonoDroid] 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2893)
[MonoDroid] 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3048)
[MonoDroid] 	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
[MonoDroid] 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
[MonoDroid] 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
[MonoDroid] 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
[MonoDroid] 	at android.os.Handler.dispatchMessage(Handler.java:106)
[MonoDroid] 	at android.os.Looper.loop(Looper.java:193)
[MonoDroid] 	at android.app.ActivityThread.main(ActivityThread.java:6669)
[MonoDroid] 	at java.lang.reflect.Method.invoke(Native Method)
[MonoDroid] 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
[MonoDroid] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
[MonoDroid] 
[Process] Sending signal. PID: 26685 SIG: 9
```

**Comment by jpobst** [comment]
The Java library needs a Java version of `checkNotNullFromProvides`. Providing a C# version of it will not work.

I don't know enough depth in Java to know what is possible. You could make a `.jar` that just has that method and include it as an `<EmbeddedJar>`, but I suspect there will be conflicts between the 2 `Preconditions` classes.

In .NET you could use some sort of assembly manipulation/rewriter like Mono.Cecil to add the method to the already compiled assembly.  I don't know if Java has something similar.

In short, you will have to solve this somehow in the Java ecosystem. I don't think there's anything in Android Bindings that can help this scenario.

**Comment by LeoJHarris** [comment]
@jpobst thanks for the response and suggestions, bugger not sure what to do on this one. 

The problem here is `Xamarin.Firebase.Messaging` which has unfortunately introduced another version of dagger i.e. `2.27.0.jar` but my binding library requires the latest dagger version i.e. 2.30.1.jar which has the method I need. So when I run my android project with both references I was getting an exception due to multiple reference's to dagger, the fix I made was I renamed the `default namespace` for my Dagger Binding library to `Xamarin.Google.Dagger` so only one Xamarin.Google.Dagger.dll was generated in my project `obj`  so was hoping I could just implement the missing method in the binding library as previously mentioned.

What about pulling the entire source of the `Xamarin.Firebase.Messaging` and putting in the latest dagger so it uses `2.30.1.jar` instead?
